storagemarket/deal_handler.go

```diff
@@ -6,20 +6,137 @@ import (
 	"fmt"
 	"sync"
 
-	"go.uber.org/atomic"
-
 	"github.com/filecoin-project/boost/storagemarket/types"
+	smtypes "github.com/filecoin-project/boost/storagemarket/types"
+	"github.com/filecoin-project/boost/storagemarket/types/dealcheckpoints"
+	"github.com/google/uuid"
 	"github.com/libp2p/go-eventbus"
 	"github.com/libp2p/go-libp2p-core/event"
-
-	"github.com/google/uuid"
+	"go.uber.org/atomic"
 )
 
+func (p *Provider) mkAndInsertDealHandler(dealUuid uuid.UUID) (*dealHandler, error) {
+	p.dhsMu.Lock()
+	defer p.dhsMu.Unlock()
+	return p.unlkMkAndInsertDealHandler(dealUuid)
+}
+
+func (p *Provider) unlkMkAndInsertDealHandler(dealUuid uuid.UUID) (*dealHandler, error) {
+	dh, ok := p.dhs[dealUuid]
+	if ok {
+		return dh, nil
+	}
+
+	dh, err := newDealHandler(p.ctx, dealUuid)
+	if err != nil {
+		return nil, fmt.Errorf("creating deal handler: %w", err)
+	}
+
+	p.dhs[dealUuid] = dh
+	return dh, nil
+}
+
+func (p *Provider) getDealHandler(id uuid.UUID) *dealHandler {
+	p.dhsMu.RLock()
+	defer p.dhsMu.RUnlock()
+
+	return p.dhs[id]
+}
+
+func (p *Provider) delDealHandler(dealUuid uuid.UUID) {
+	p.dhsMu.Lock()
+	delete(p.dhs, dealUuid)
+	p.dhsMu.Unlock()
+}
+
+// startDealThread sets up a deal handler and wait group monitoring for a deal, then
+// executes the deal in a new go routine
+func (p *Provider) startDealThread(deal *types.ProviderDealState) (bool, error) {
+	p.dhsMu.Lock()
+	defer p.dhsMu.Unlock()
+
+	// Check if the deal is already running
+	_, alreadyRunning := p.dealThreads[deal.DealUuid]
+	if alreadyRunning {
+		return false, nil
+	}
+	p.dealThreads[deal.DealUuid] = struct{}{}
+
+	// Set up deal handler so that clients can subscribe to deal update events
+	dh, err := p.unlkMkAndInsertDealHandler(deal.DealUuid)
+	if err != nil {
+		return false, fmt.Errorf("creating deal handler: %w", err)
+	}
+
+	p.runWG.Add(1)
+	go func() {
+		defer p.runWG.Done()
+		defer func() {
+			p.dhsMu.Lock()
+			delete(p.dealThreads, deal.DealUuid)
+			p.dhsMu.Unlock()
+		}()
+
+		// Run deal
+		p.runDeal(deal, dh)
+		p.dealLogger.Infow(deal.DealUuid, "deal go-routine finished execution")
+	}()
+
+	return true, nil
+}
+
+func (p *Provider) failPausedDeal(deal *smtypes.ProviderDealState) error {
+	p.dhsMu.RLock()
+	defer p.dhsMu.RUnlock()
+
+	// Check if the deal is running
+	_, alreadyRunning := p.dealThreads[deal.DealUuid]
+	if alreadyRunning {
+		return fmt.Errorf("the deal %s is running; cannot fail running deal", deal.DealUuid)
+	}
+
+	dh, ok := p.dhs[deal.DealUuid]
+	if !ok {
+		// This should never happen, but check just in case
+		return fmt.Errorf("the deal %s does not have a deal handler; cannot fail deal", deal.DealUuid)
+	}
+
+	// Update state in DB with error
+	deal.Checkpoint = dealcheckpoints.Complete
+	deal.Retry = smtypes.DealRetryFatal
+	errMsg := "user manually terminated the deal"
+	err := errors.New(errMsg)
+	if deal.Err != "" {
+		err = errors.New(deal.Err)
+	}
+	deal.Err = errMsg
```

Comment:
Nit, but I find this bit a bit confusing...

Maybe have something like:

```suggestion
	var err error
	if deal.Err != "" {
	    err = errors.New(deal.Err)
	} else {
	    err = errors.New("user manually terminated the deal")	
	}
	deal.Err = "user manually terminated the deal"
```